### PR TITLE
Strip units on passed in values to em/rem functions & mixins. Fixes #80

### LINKED ISSUE #80: Strip units on passed in values to em & rem functions and mixins
Currently, the `(to)-em` and `(to)-rem` mixins and functions don't handle anything but unitless values.

I'm so used to passing in a `px` on the end of my string that the output then breaks. It would be great to strip units on whatever was supplied before the calculation.

### PR DESCRIPTION
**Pull request by stowball**
I've implemented `strip-units` to make these functions and mixins more robust. It also makes it less annoying when you have to change from raw px values to a function or mixin.

As v2 of Scally is no longer using `rem`s on things like `border` and `box-shadow`, it makes it inconsistent that for these properties you have to type `px` values, but for `em` and `rem` functions, you have to use unitless.

At CM, I encountered this where I set variables for `width`s, and then @chris-pearce asked me to convert them to `rem`s. At first I just wrapped them in a `to-rem()` function, but that errored, so had to change my values.

With this PR, a user can mix and match or use whatever they prefer.

### CODE REVIEW

File: style.scss [hunk 1]
```diff
@@ -114,6 +114,8 @@
 
 @import "bower_components/scally/core/functions/string-replace";
 
+@import "core/functions/strip-unit";
```
Comment (chris-pearce) [inline style.scss]: I know this isn't the best setup and this file will get blown away from the repo but for the sake of consistency can we follow the format of the all the other `@import` paths?

**Comment by chris-pearce** [comment]
Thanks for doing this but I need to look into this more. And off the top of my head; what about other mixins that except unitless values? Like:
- https://github.com/chris-pearce/scally/blob/master/core/mixins/_font-size.scss
- https://github.com/chris-pearce/scally/blob/master/core/mixins/_media-queries.scss

Plus GH saying this can't be merged, probably `style.css`?

File: style.scss (hunk 1 above)
Comment (stowball) [inline style.scss]: Ah crap, sorry. It was at this point that I noticed th e font bug and got distracted, and forgot to fix the path :(

**Comment by stowball** [comment]
Yeah, they should also be changed. I haven't used them yet, which is why they never crossed my mind :(

And yeah, style.CSS is a pain in the ass

**Comment by chris-pearce** [comment]
What about setting values which are all unitless e.g. 

```
$font-size: 16;
```

Well they were but now V2 has some `px` units e.g. https://github.com/chris-pearce/scally/blob/v2/objects/_o-arrow.scss#L22 which I agree is inconsistent so in these cases maybe I have to apply the unit within Scally to keep those `px` settings unitless e.g.

```
border: $o-arrow-size-base * 1px solid transparent;
```

https://github.com/chris-pearce/scally/blob/v2/objects/_o-arrow.scss#L41

**Comment by chris-pearce** [comment]
> > Yeah, they should also be changed. I haven't used them yet, which is why they never crossed my mind :(

Yeah it would have to be applied across the board.

> > And yeah, style.CSS is a pain in the ass

I know, don't commit it? I'd remove it now but I have to look into the package json setups as those can't break.

**Comment by stowball** [comment]
That's a crazy amount of overhead. And there's no way that anyone writing their custom CSS would adhere to that, so then you've instantly got 2 different styles of authoring units in a project

**Comment by chris-pearce** [comment]
> > That's a crazy amount of overhead. And there's no way that anyone writing their custom CSS would adhere to that, so then you've instantly got 2 different styles of authoring units in a project

What are you referring to exactly? Please reference previous comments via ">>".

**Comment by stowball** [comment]
> > What are you referring to exactly? Please reference previous comments via ">>".
> > `border: $o-arrow-size-base * 1px solid transparent;`

(Quoting is hard work on a phone)

**Comment by stowball** [comment]
> > What are you referring to exactly? Please reference previous comments via ">>".
> > `border: $o-arrow-size-base * 1px solid transparent;`

(Quoting is hard work on a phone)

**Comment by chris-pearce** [comment]
```
border: $o-arrow-size-base * 1px solid transparent;
```

I'm talking about handling this **in the framework** not expecting Scally users to do this and I'm doing this so Scally users can continue to apply unitless values to Scally settings e.g.

```
$o-arrow-size-base: 8;
```

To keep it consistent with the majority of other Scally settings which take `rem` units and up until V2 **ALL** Scally settings only accept unitless values.

Some examples:

```
$font-size: 16  !default;

$border-radius: 4 !default;

$breakpoints: (

  // Palm
  "palm":         640,

  // Lap
  "lap":          641,
  "lap-large":    901,

  // Desk
  "desk":         1200,
  "desk-large":   1366

) !default;

$text-input-and-select-border-thickness: 1 !default;

$hr-thickness: 1 !default;
```

I think it's confusing for the few Scally **Version 2** settings that expect `px` now (border, box-shadow, etc) to have to apply the `px` unit. I want it all consistent as for sure after applying unitless values to the majority of Scally settings people will expect every Setting to work that way.

**Comment by stowball** [comment]
As I mentioned in my original issue, I personally prefer passing in px values to things, not unitless. I want to tell the mixin to use 20px, that's how my brain works, as I'm sure most web developers' brains do. 20px is my intended target (however that gets transformed on the other side), not just 20.

I'd be all for changing all values in Scally to have px on them.

**Comment by kllevin** [comment]
I disagree, I think Scally should continue to be unitless. It is expected that any framework will have certain conventions you need to follow in order for the framework to function. While the goal of any framework developer should be to keep the learning curve low, I don't think it is unreasonable to expect a Scally user to have to learn a few Scally specific conventions. In this case it's learning to pass unitless values. In a responsive world passing 20px doesn't mean much anyway.

**Comment by chris-pearce** [comment]
> > I personally prefer passing in px values to things, not unitless

I'm the opposite. And when Scally was created it was a very conscious decision to not make users of the fw apply units. 

> > that's how my brain work

Not mine :) 

I'll expand on this later as I need to crack on with V2.

**Comment by chris-pearce** [comment]
Wow mine and @kllevin reply happened same time.

**Comment by stowball** [comment]
> > In a responsive world passing 20px doesn't mean much anyway.

I just don't get this. How is 20px any different to 20? The output will be identical.

Anyway, it was just a possible solution to mixed units issue. And, with stripping in place, you can quite happily continue passing in unitless if you like.

**Comment by stowball** [comment]
OK, I've stripped units on media-queries and font-size.

But I've also done something controversial: If you pass in a px `line-height` to the `font-size()` mixin, it will convert it to a unitless value.

So `font-size(12[px], 18px)` will output `line-height: 1.5`

IMO this is one of the strengths of using a pre-processor like Sass. I don't want to have to manually calculate that the `line-height` I need is 1.5, but I do know (as seen in the Photoshop comp) that it is 18px.

**Comment by chriswrightdesign** [comment]
Long time listener, first time poster - just catching up on v2

In regards to the unitless and units discussion for passing arguments into functions, I think there’s a definite argument for maintainability to show the unit that people are passing in. Whether your way of thinking is pixels, ems, or rems - if another developer less well versed in the framework were to read your code, would it be obvious to them as to what was actually happening? What’s obvious to the creators of something isn’t always obvious to the users, being explicit can only be helpful for new adoption of the framework, requires less teaching, and causes less frustration. 

I personally think in ems (thanks to years of multiplying by 16), others think in pixels - it’d be nice to know what figure went in. It does seem like Scally is using pixels anyway just without the unit. 

In regards to there being a forced unit (rem) by convention, it seems contrary to what Scally is advertised as - an unopinionated framework. We have a toolbox in front of us, it seems weird to say we’re only going to use the hammer for everything, when a screwdriver might do the job better. Rems are a useful and powerful unit, but they’re not the only useful and powerful unit. 

We have different unit types in CSS for a lot of reasons, some are scalable (ems), some are precise (px), there are plenty use cases for other unit types: VH - Full screen without CSS hacks or JS, VW - full screen, type , % - flexbox fallback, pixels - borders/ dropcaps / flexbox images.

In typography for example, there’s a strength in a typographic system being relative to its module over being relative to the document root, and it can be useful and more flexible to allow its font-sizing to cascade through the module. In this way ems are more flexible than rems. As Jason Santa Maria, who literally wrote the book On Web Typography (highly recommend reading!), says "Type is such a finicky beast that it needs to be judged optically first, not by the numbers.” 

Convention and consistency are fine, but unit types shouldn’t be where the convention is enforced - units are different sets of tools applied to different situations.

Anyway, my two cents.

**Comment by chris-pearce** [comment]
I can see the arguments for this making the framework more robust and flexible and I’m all for making the framework less frustrating. Up until now there hasn’t been any issues around not being able to attach “px” after the number you pass into mixins such as `to-rem` and `to-em` however after giving this more thought I do see how this can help the framework. 

As this is a non breaking update to the framework I’m happy to add this, however I’m still not  comfortable everything has been covered. I’d like to discuss the changes more in general and so I will catch up with @stowball in person to go over it.

To echo what @kllevin said, you don't commit to using a CSS framework without spending some time learning its more complicated parts such as its mixins and functions i.e. its tools. I believe one of Scally’s strengths is everything is heavily commented [e.g.](https://github.com/chris-pearce/scally/blob/master/core/mixins/_convert-px-to-em-rem.scss#L7-L21) and it has fully fleshed out README’s. This should also aid new adopters of the framework in understanding exactly what is going on.

The areas of the framework we’re specifically talking about effecting in relation to this PR are:

**Functions**
- [convert-px-to-em-rem.scs](https://github.com/chris-pearce/scally/blob/master/core/functions/_convert-px-to-em-rem.scss)
  
  ^ a specific tool that takes a number and converts it to `rem` or `em`.

**Mixins**
- [convert-px-to-em-rem.scss](https://github.com/chris-pearce/scally/blob/master/core/mixins/_convert-px-to-em-rem.scss)
  
  ^ see above.
- [font-size](https://github.com/chris-pearce/scally/blob/master/core/mixins/_font-size.scss)
  
  ^ a specific tool that takes a number and converts it to `rem` and generates an optimal unitless `line-height`.
- [media-queries](https://github.com/chris-pearce/scally/blob/master/core/mixins/_media-queries.scss)
  
  ^ a specific tool that takes a number and converts it to `em`.

**Settings**
- [Breakpoints](https://github.com/chris-pearce/scally/blob/master/core/settings/_breakpoints.scss)
- [Spacing](https://github.com/chris-pearce/scally/blob/master/core/settings/_spacing.scss)
- [Typography](https://github.com/chris-pearce/scally/blob/master/core/settings/_typography.scss)

---

Most frameworks—ones that I’m familiar with anyway—choose a **core** unit, some examples:
- [**Inuit**](https://github.com/inuitcss) chooses `px` e.g.
  
  https://github.com/inuitcss/settings.defaults/blob/master/_settings.defaults.scss#L19.
- [**SUIT**](https://github.com/suitcss) doesn’t really have one but in places enforces a unit like `px` e.g.
  
  https://github.com/suitcss/components-grid/blob/master/lib/grid.css#L5 
  https://github.com/suitcss/components-grid/blob/master/lib/grid.css#L4 
- [**Bootstrap**](https://github.com/twbs/bootstrap) chooses `px` e.g.  
  
  https://github.com/twbs/bootstrap/blob/master/less/variables.less#L51
  https://github.com/twbs/bootstrap/blob/master/less/variables.less#L90-L100, etc. 
- [**UI Kit**](https://github.com/uikit/uikit/) chooses `px` e.g.
  
  https://github.com/uikit/uikit/blob/master/src/less/core/base.less#L17 

I guess you can strive to create a framework—that at its core—supports more than just one unit e.g.
- Offering the **core** typography system for font size to work with `px` and `em`, not just `rem`?
  
  https://github.com/chris-pearce/scally/blob/master/core/settings/_typography.scss#L11
  https://github.com/chris-pearce/scally/blob/master/core/base/_root.scss#L30
- Offering media queries to work with `px`, not just `em`?
  
  https://github.com/chris-pearce/scally/blob/master/core/settings/_breakpoints.scss#L44-L57 
  https://github.com/chris-pearce/scally/blob/master/core/mixins/_media-queries.scss
- Offering the spacing settings and utility to work with `px` and `em` not just `rem`?
  
  https://github.com/chris-pearce/scally/blob/master/core/settings/_spacing.scss 
  https://github.com/chris-pearce/scally/blob/master/utilities/_u-spacing.scss 

This is trying to do too much IMO.

Apart from the above Scally does not strictly enforce any other unit, for example, if in your component you want to use `em` for padding then you can do so, and is something we do at CM, e.g.

``` scss
.c-pagination {
  @include to-em(padding, 24);
}
```

It’s worth saying though that perhaps Scally’s [**Button**](https://github.com/chris-pearce/scally/blob/master/components/_c-button.scss) component needs to be revisited as the padding is using `rem`, [see](https://github.com/chris-pearce/scally/blob/master/components/_c-button.scss#L141) where it should perhaps be `em`? Maybe you can counter that with “but `em`s are too opinionated.” I guess you can’t please every scenario.

There will always be certain parts of a framework that have to be a little opinionated. In my opinion Scally does a great job minimising its opinionated bits and definitely does not force you to use `rem` for everything.

I had a little scout around and found some examples of other people's mixins/functions that **do** accept unitless _and_ applying "px" and some examples that **don't** accept units as per Scally.
- Bourbon
  - http://bourbon.io/docs/#px-to-em
  - http://bourbon.io/docs/#px-to-rem
- SCUT
  - http://davidtheclark.github.io/scut/em.html
  - http://davidtheclark.github.io/scut/rem.html  
  - http://davidtheclark.github.io/scut/rem-fallback.html 
- https://css-tricks.com/snippets/css/less-mixin-for-rem-font-sizing/
- https://github.com/justinskolnick/unitize 
- https://github.com/zurb/foundation/blob/master/scss/foundation/_functions.scss#L106-L135 
- https://web-design-weekly.com/2013/05/12/handy-sass-mixins/ -> Font size

**Comment by stowball** [comment]
@chris-pearce As discussed, I've converted all settings that use px to be unitless, and am stripping the px inside the property. This means that by default, all Scally values are now unitless, but an author could use unitless or px throughout a Scally project whereever they like

**Comment by chris-pearce** [comment]
@stowball nice work, looks good. I'll pull it down tomorrow and have a test, I don't trust myself now as I'm knackered. You'll need to rebase master in order to get the green merge button :+1:.

**Comment by chris-pearce** [comment]
@kllevin have a gander.

**Comment by stowball** [comment]
I had the green button yesterday! Damn you LibSass fix! :P

All merged and updated now.

**Comment by stowball** [comment]
Also, as discussed, it's worth noting that 90% of the links to other frameworks that @chris-pearce provided strip units so accept both px and unitless. It'll be great to have Scally inline with the "general population" to make adoption easier.

**Comment by chris-pearce** [comment]
Yes providing the ability to pass in `px` and a raw number makes sense. I just wanted to highlight that not everyone needs the `px` bit :). 

And making sure it's understood that frameworks need to work with a _core_ unit for some parts of the framework and that Scally holds true to being as _unopinionated_ as it can possibly be esp, in the _zero design_ department.

I'll get it merged tonight.

**Comment by chris-pearce** [comment]
Closing this as its moved to PR #124.

**Pull request closed without merging**